proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+### `ContiguousStorage`
+
+A type can indicate that it can provide a `Span` by conforming to the `ContiguousStorage` protocol. `ContiguousStorage` forms a bridge between multi-type or generically-typed interfaces and a performant concrete implementation.
+
+For example, for the hypothetical base64 decoding library mentioned above, a possible API could be:
+
+```swift
+extension HypotheticalBase64Decoder {
+  public func decode(bytes: some ContiguousStorage<UInt8>) -> [UInt8]
+}
+```
+
+Even better, an interface can be defined in terms of the concrete type `Span<UInt8>`:
+
+```swift
+extension Hypothetical Base64Decoder {
+  public func decode(bytes: Span<UInt8>) -> [UInt8]
+}
+```
+
+Advanced libraries might add use an inlinable generic-dispatch interface in addition to a concrete interface defined in terms of `Span`
+
+### `RawSpan`
+
+`RawSpan` allows sharing the contiguous internal representation for values which may be heterogenously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straight-forwards understanding of performance for library code.
+
+`Span<some BitwiseCopyable>` can always be converted to `RawSpan`, using a conditionally-available property or a constructor.
+
+## Detailed design
+
+`Span<Element>` is a simple representation of a span of initialized memory.
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+It provides a buffer-like interface to the elements stored in that span of memory:
+
+```swift
+extension Span {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+  public var indices: Range<Int> { get }
+
+  subscript(_ position: Int) -> Element { get }
+}
+```
+
+Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the view from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.
+
+`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the recently added `extracting()` functions of `UnsafeBufferPointer`:
+
+```swift
+extension Span {
+  public func extracting(_ bounds: Range<Int>) -> Self
+}
+```
+
```

Comment:
To me, subjectively, subscript reads better.  SE-0437's justification was because we don't have `Slice<~Copyable>`, but that doesn't apply if we don't have a slice type. 

Perhaps our rationale is that we want to reserve the right to a different slice type in the future.

Reply:
I agree that subscript reads better, but the stdlib's pattern is that what's returned by the range subscript getter shares its indices with the original binding. It feels wrong to make something that behaves so differently while looking the same. Discussions welcome.

Reply:
The slicing subscript notation would indeed read better. However, Swift has very strongly established the expectation that the result of `foo[i ..< j]` will share indices with the original container, which is not the case here. We should not use the same notation to mean a very different thing; especially not in the context of a shape that is so close to our preexisting primitive container types.

Additionally, the slicing subscript notation we have in classic Swift is able to accommodate all three flavors of access (borrowing, mutating and consuming) with a single, universal result type. This isn't going to be possible within the ownership model -- we'll need to invent distinct notations for extracting borrowing vs mutating parts. I don't think `Span` is the right place to establish a general pattern for this matter.